scatter_gather.go

```diff
@@ -29,21 +29,22 @@ func (sg *scatterGather[T, R]) scatter(ctx context.Context, forEach func(T) (<-c
 			default:
 			}
 
-			sout, err := forEach(target)
+			cctx, cncl := context.WithTimeout(ctx, sg.maxWait)
+			sout, err := forEach(cctx, target)
+			cncl()
 			if err != nil {
 				log.Errorw("failed to scatter on target", "target", target, "err", err)
 				return
 			}
 
-			for {
-				select {
-				case <-ctx.Done():
-					return
-				case r, ok := <-sout:
-					if !ok {
+			if sout != nil {
```

Comment:
Missing nil check might have been the root cause of timeouts in http? I think selecting on nil channel blocks forever?

Reply:
i thought it returned immediately not okay?

Reply:
This hangs for me: 

```go
func TestChannel(t *testing.T) {
	var c <-chan string = nil
	select {
	case <-c:
		t.Log("selected")
	}
	t.Log("done")
}
```

Reply:
looks like it!

Reply:
The non-intuitive things is `case ss, ok := <-c:` also blocks.